### PR TITLE
feat: add realtime-transcribe mode for OpenClaw and voice mode for Hermes

### PR DESCRIPTION
**Pull request by Ananaskelly**
## Summary

Extended OpenClaw voice support with transcribe mode and implemented voice mode support for Hermes.

## Implementation Notes :hammer_and_pick:

For OpenClaw: a new voice-transcribe mode was added. This mode also streams audio chunks in a separate session, but unlike realtime mode, it only generates a transcript and returns it as the result. The agent is then run with the transcribed message, similarly to the default text mode.

For Hermes: in actual use, a voice message is recorded through the CLI, passed to the ASR engine as a whole, and then added to the ongoing agent session. Similarly, in this implementation, the synthesized audio is sent to the ASR engine through the Hermes API, and the agent is then run with the transcribed message.

## External Dependencies :four_leaf_clover:

* 

## Breaking API Changes :warning:

* 

## Issues that this PR closes

* 

*Simply specify none (N/A) if not applicable.*

## Summary by CodeRabbit

- **New Features**
  - Added voice-transcription mode for OpenClaw and Hermes.
  - Supports speech-to-text input, text-to-speech output, transcript-driven agent execution, and saved voice artifacts.
  - Added configuration for OpenAI and local speech providers, including models, language, voice, speed, and timeouts.
  - Added TerminalBench2 voice-transcription profiles for OpenClaw and Hermes.

- **Bug Fixes**
  - Improved transcription session handling, validation, credential protection, cleanup, and error reporting.
  - Preserved existing agent and realtime behavior while adding voice-transcription support.

- **Documentation**
  - Added guidance for configuring voice modes and understanding generated artifacts.

### CODE REVIEW

File: profiles/hermes-tb2-fix-git-voice-transcribe-deepseek.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  "benchmark": {
+    "type": "terminalbench2",
+    "root": ".cache/terminal-bench-2",
+    "tasks": [
+      "fix-git"
+    ]
+  },
+  "harness": {
+    "type": "hermes",
+    "mode": "voice-transcribe",
```
Comment (leokondrashov) [inline profiles/hermes-tb2-fix-git-voice-transcribe-deepseek.json:17]: i would call those the same way for both openclaw and heremes in the configs for consistency. those are the same process under the hood, so better to have the same name here, even though the harnesses call those differently.

File: pkg/harness/openclaw/harness.go [hunk 2]
```diff
@@ -60,19 +60,24 @@ exit "$status"`
 var gatewayPort = network.MustParsePort(gatewayListenPort + "/tcp")
 
 const (
-	ModeAgent    = "agent"
-	ModeRealtime = "realtime"
+	ModeAgent              = "agent"
+	ModeRealtimeTalk       = "realtime-talk"
+	ModeRealtimeTranscribe = "realtime-transcribe"
```
Comment (leokondrashov) [inline pkg/harness/openclaw/harness.go]: does it make sense to name them after realtime? because transcribe doesn't seem to be the realtime, it very sequential with the steps, while the talk mode is actual realtime (extra agent the listens all the time)

**Review by leokondrashov — COMMENTED** [review]
please also add the docs for the voice modes, added or already existing. we need to be clear what actually each mode is doing so we know where the caveats might lie in perf/accuracy.

File: pkg/harness/openclaw/harness.go (hunk 2 above)
Comment (Ananaskelly) [reply pkg/harness/openclaw/harness.go]: Yes, I understand. I was also thinking about the naming. Initially, it came from the openclaw docs at the time (they use a different name now). Since hermes uses a different way of delivering audio, I left it as is. But I agree that using the same "voice-transcribe" name might be more obvious. I'll change it.

File: pkg/harness/hermes/harness.go [hunk 3]
```diff
@@ -459,6 +564,274 @@ func (manager *Manager) Run(ctx context.Context, instruction string) (core.Harne
...
+	RawTranscript string         `json:"raw_transcript"`
+	ReturnCode    int            `json:"returncode"`
+	RawResult     map[string]any `json:"raw_result,omitempty"`
+	Stdout        string         `json:"stdout"`
+	Stderr        string         `json:"stderr"`
+	Signature     string         `json:"signature,omitempty"`
+	Error         string         `json:"error,omitempty"`
+}
+
+func (manager *Manager) synthesizeVoiceInstruction(ctx context.Context, active *session, instruction string) (string, []string, error) {
```
Comment (leokondrashov) [inline pkg/harness/hermes/harness.go:636]: this seems to be very similar to synthesizeRealtimeAudio in openclaw harness, can it be deduplicated, at least partially?

File: docs/voice_mode.md
Comment (leokondrashov) [inline docs/voice_mode.md:1]: include a common comment about the timeout resets for the stt mode. we should explain that we give separate timeout for the stt part of execution and actual task session gets the same amount of time to finish the task.

File: pkg/config/config.go [hunk 4]
```diff
@@ -618,60 +632,103 @@ func (h *HarnessConfig) validate() error {
...
-		}
-		if h.Realtime.QuietDuration, err = parseOptionalPositiveDuration("harness.realtime.quiet_duration", h.Realtime.QuietDurationText); err != nil {
-			return err
-		}
-		if h.Realtime.AgentWaitDuration, err = parseOptionalPositiveDuration("harness.realtime.agent_wait_duration", h.Realtime.AgentWaitDurationText); err != nil {
-			return err
-		}
-		if h.Realtime.ToolCallTimeout, err = parseOptionalPositiveDuration("harness.realtime.tool_call_timeout", h.Realtime.ToolCallTimeoutText); err != nil {
-			return err
+		return h.Realtime.validate()
```
[the remark this answers is not in the document]
Comment (leokondrashov) [reply pkg/config/config.go:675]: this seems to be divergence from the pr description rather than the code issue. can you update the description so we wouldn't diverge?

**Review by leokondrashov — COMMENTED** [review]
please update the design/harness.md to cover the newly added voice mode. maybe update the openclaw's section as well.

**Comment by leokondrashov** [comment]
please also update the wordlist.txt to include some acronims used in docs so the linters would pass

File: cmd/aries/wiring.go [hunk 5]
```diff
@@ -222,26 +222,7 @@ func environmentFromConfig(cfg *config.BenchmarkEnvironment) core.Environment {
...
-			AgentWaitDuration:     cfg.Harness.Realtime.AgentWaitDuration,
-			ToolCallTimeout:       cfg.Harness.Realtime.ToolCallTimeout,
-			TrailingSilenceMillis: cfg.Harness.Realtime.TrailingSilenceMillis,
-			Provider:              cfg.Harness.Realtime.Provider,
-			Model:                 cfg.Harness.Realtime.Model,
-			Voice:                 cfg.Harness.Realtime.Voice,
-			ReasoningEffort:       cfg.Harness.Realtime.ReasoningEffort,
-			IncludeEvents:         cfg.Harness.Realtime.IncludeEvents,
-		}
+		realtime := openClawRealtimeOptions(cfg.Harness)
```
Comment (chengzhi-lu) [inline cmd/aries/wiring.go]: This voice-mode initialization feels a bit too implementation-specific for `newHarness()`. It now needs to understand and manually translate Hermes TTS/STT details, and `VoiceTranscribe` is populated regardless of whether that mode is actually active. Could we move this into a OpenClaw/Hermes-specific helper/config path and only construct the voice options when the corresponding mode is enabled? 
For example 
```
if cfg.Harness.Mode == "voice-transcribe" {
    voice = openClawRealtimeOptions(cfg.Harness)
}
```

File: pkg/harness/openclaw/harness.go [hunk 6]
```diff
@@ -634,6 +647,34 @@ func (manager *Manager) runRealtime(ctx context.Context, active *session, instru
 	}, nil
 }
 
+func (manager *Manager) runAgentWithTranscript(ctx context.Context, active *session, client gatewayConnection, result *realtimeclient.Result) error {
+	transcript := strings.TrimSpace(result.Transcript)
```
Comment (chengzhi-lu) [inline pkg/harness/openclaw/harness.go:670]: This function runs once after the entire task audio has been streamed and the transcription runner has finished. However, a single recording can produce multiple transcript.done events when pauses split the speech into separate turns. Each event overwrites result.Transcript (see openclaw/realtime/runner.go:420-426), so this function may receive only the last segment and omit earlier task requirements.

Please assemble the final segments from this recording into the complete agent input. A test with two final segments should verify that both reach the agent in order.

File: pkg/harness/openclaw/harness.go [hunk 7]
```diff
@@ -610,6 +620,9 @@ func (manager *Manager) runRealtime(ctx context.Context, active *session, instru
 	}
 	runCtx, cancel := context.WithTimeout(ctx, active.agentTimeout)
 	realtimeResult, err := runner.Run(runCtx)
+	if err == nil && manager.mode == ModeVoiceTranscribe {
+		err = manager.runAgentWithTranscript(runCtx, active, client, &realtimeResult)
```
Comment (chengzhi-lu) [inline pkg/harness/openclaw/harness.go]: The transcription runner exits before client.Agent() starts on the same delivery-mode connection. Agent() only consumes responses, leaving events queued. Once the queue exceeds 2048 events or 16 MiB, the client terminates the connection with gateway event queue overflow, potentially losing the outcome of an accepted task. Please continue draining events or safely switch to response-only handling, and test this with the actual Gateway client receiving sustained events.

File: pkg/harness/hermes/harness.go [hunk 8]
```diff
@@ -433,6 +535,9 @@ func (manager *Manager) Run(ctx context.Context, instruction string) (core.Harne
 	active = &local
 	manager.mu.Unlock()
 
+	if manager.mode == ModeVoiceTranscribe {
+		return manager.runVoiceTranscribe(ctx, active, instruction, started)
+	}
 	runCtx, cancel := context.WithTimeout(ctx, active.agentTimeout)
```
Comment (chengzhi-lu) [inline pkg/harness/hermes/harness.go:541]: The agent timeout has different semantics across the two harnesses. OpenClaw starts active.agentTimeout before transcription, so audio streaming, transcription, and the subsequent agent run share that budget. Hermes starts the agent timeout after STT completes, giving the agent the full budget.

With the same timeout setting, OpenClaw therefore leaves less time for agent execution as audio duration or transcription latency increases, which can affect comparisons of task success rates. Could we use consistent timeout boundaries across both harnesses, with separate transcription and agent budgets or a shared end-to-end budget?

File: pkg/harness/openclaw/realtime/runner.go [hunk 9]
```diff
@@ -339,6 +382,9 @@ func (runner *Runner) processEvents(ctx context.Context, result *Result) error {
 	} else {
 		result.Transcript = state.latestUserTranscript
 	}
+	if runner.options.SessionMode == SessionModeTranscribe && result.TranscriptDone == "" {
```
Comment (chengzhi-lu) [inline pkg/harness/openclaw/realtime/runner.go]: A non-empty TranscriptDone only confirms that one segment finished. If a later segment remains partial when the quiet timeout expires, this check still passes and the agent receives incomplete input. Could we verify that the full recording has been transcribed before starting the agent?

File: cmd/aries/wiring.go (hunk 5 above)
Comment (Ananaskelly) [reply cmd/aries/wiring.go]: https://github.com/hyscale-lab/ARIES/blob/71308be005c4dab83c546be4b2980f8f99604c1a/cmd/aries/wiring.go#L232

https://github.com/hyscale-lab/ARIES/blob/71308be005c4dab83c546be4b2980f8f99604c1a/cmd/aries/wiring.go#L248

Moved voice options for both Hermes and OpenClaw into separate configs and made them constructed only when needed. Renamed the corresponding configs to *hermesVoiceOptions* and *openClawVoiceOptions* for consistency.

File: pkg/harness/openclaw/harness.go (hunk 6 above)
Comment (Ananaskelly) [reply pkg/harness/openclaw/harness.go:670]: https://github.com/hyscale-lab/ARIES/blob/afad319b4fc3f47d3c9d26f79644a11671fdf9a6/pkg/harness/openclaw/harness.go#L671

Fixed to use all joined transcript.done segments.

File: pkg/harness/openclaw/harness.go (hunk 7 above)
Comment (Ananaskelly) [reply pkg/harness/openclaw/harness.go]: https://github.com/hyscale-lab/ARIES/blob/afad319b4fc3f47d3c9d26f79644a11671fdf9a6/pkg/harness/openclaw/harness.go#L630

Fixed by closing the realtime gateway connection after the transcribe session and opening a new response-only gateway connection for the agent run.

File: pkg/harness/hermes/harness.go (hunk 8 above)
Comment (Ananaskelly) [reply pkg/harness/hermes/harness.go:541]: https://github.com/hyscale-lab/ARIES/blob/afad319b4fc3f47d3c9d26f79644a11671fdf9a6/pkg/harness/openclaw/harness.go#L625

Fixed by adding a separate STT timeout for OpenClaw voice-transcribe mode and starting the agent timeout only after transcription completes. For realtime mode, I don't have a good idea of how to make it consistent with voice-transcribe mode. We could extend the timeout, but in that case the agent would most likely have more time to complete the task.

File: pkg/harness/openclaw/realtime/runner.go (hunk 9 above)
Comment (Ananaskelly) [reply pkg/harness/openclaw/realtime/runner.go]: https://github.com/hyscale-lab/ARIES/blob/afad319b4fc3f47d3c9d26f79644a11671fdf9a6/pkg/harness/openclaw/realtime/runner.go#L389

Fixed by adding a new pending state for transcript.delta events and checking whether the transcript is incomplete before running the agent.

File: pkg/harness/hermes/harness.go [hunk 10]
```diff
@@ -459,6 +564,248 @@ func (manager *Manager) Run(ctx context.Context, instruction string) (core.Harne
...
+		err = redactSessionError(err, active)
+		return failedHarnessResult(active, started, err), err
+	}
+	if strings.TrimSpace(sttResult.Transcript) == "" || strings.ContainsRune(sttResult.Transcript, 0) {
+		err := redactSessionError(errors.New("Hermes voice transcript is invalid"), active)
+		return failedHarnessResult(active, started, err), err
+	}
+	runCtx, cancel := context.WithTimeout(ctx, active.agentTimeout)
+	result, runErr := manager.execAttached(runCtx, active.containerID,
+		[]string{agentWrapperPath, active.model.Model, active.model.Provider, sttResult.Transcript}, workspaceRoot)
```
Comment (chengzhi-lu) [inline pkg/harness/hermes/harness.go]: Please use `hermesProvider(active.model.Provider)` rather than directly pass the `active.model.Provider `here, as the text path does. For openai and sglang, the rendered config uses custom, but this invocation passes the original backend name to --provider, which the pinned Hermes version does not support. Transcription can therefore succeed while agent startup fails. Please add voice-mode tests checking the provider argument for both backends.

File: pkg/harness/openclaw/harness.go [hunk 11]
```diff
@@ -601,16 +613,36 @@ func (manager *Manager) runRealtime(ctx context.Context, active *session, instru
...
-	runCtx, cancel := context.WithTimeout(ctx, active.agentTimeout)
-	realtimeResult, err := runner.Run(runCtx)
-	cancel()
+	var realtimeResult realtimeclient.Result
+	if manager.mode == ModeVoiceTranscribe {
+		transcribeCtx, transcribeCancel := context.WithTimeout(ctx, defaultVoiceSTTTimeout)
+		realtimeResult, err = runner.Run(transcribeCtx)
+		transcribeCancel()
+		closeErr := client.Close()
+		if err == nil {
```
Comment (chengzhi-lu) [inline pkg/harness/openclaw/harness.go:634]: err == nil does not establish that transcription succeeded. The runner records session.error and turn.cancelled in realtimeResult.Errors without returning an error. A sequence such as transcript.done(A) -> session.error -> quiet timeout can therefore start the agent with only A; the recorded errors are checked only after execution. Please handle transcription errors before submitting the agent request and add a test asserting zero agent calls for this sequence.

File: pkg/harness/openclaw/realtime/runner.go [hunk 12]
```diff
@@ -339,6 +382,14 @@ func (runner *Runner) processEvents(ctx context.Context, result *Result) error {
 	} else {
 		result.Transcript = state.latestUserTranscript
 	}
+	if runner.options.SessionMode == SessionModeTranscribe {
+		if result.TranscriptDone == "" {
+			return errors.New("missing_final_transcript: no final transcript.done event was observed")
+		}
+		if state.partialTranscriptPending {
```
Comment (chengzhi-lu) [inline pkg/harness/openclaw/realtime/runner.go:389]: This fixes the trailing-delta case, but it does not establish that the entire recording has been transcribed. If A is finalized and B's first event arrives after the quiet timeout, this flag remains false and the runner accepts A alone. Could we use a completion signal covering all submitted audio, or document and verify the protocol guarantee that makes this timeout sufficient? Please add a delayed-final-segment test.

**Review by chengzhi-lu — COMMENTED** [review]
Please update docs/design/harness.md to describe the separate agent gateway connection, and document OpenClaw's separate STT timeout and realtime timeout semantics in docs/voice_mode.md. Also remove the leftover >>>>>>> refs/rewritten/main marker from .github/configs/wordlist.txt and align the PR description with the public mode names realtime and voice-transcribe. @Ananaskelly

File: pkg/harness/hermes/harness.go (hunk 10 above)
Comment (Ananaskelly) [reply pkg/harness/hermes/harness.go]: https://github.com/hyscale-lab/ARIES/blob/4db567993ab9e78e721569eba95bb490ad215804/pkg/harness/hermes/harness.go#L609

Changed to use hermesPovider().

https://github.com/hyscale-lab/ARIES/blob/4db567993ab9e78e721569eba95bb490ad215804/pkg/harness/hermes/harness_test.go#L692

Added the test.

File: pkg/harness/openclaw/harness.go (hunk 11 above)
Comment (Ananaskelly) [reply pkg/harness/openclaw/harness.go:634]: https://github.com/hyscale-lab/ARIES/blob/4db567993ab9e78e721569eba95bb490ad215804/pkg/harness/openclaw/harness.go#L630

Added error handling before submitting the agent request.

File: pkg/harness/openclaw/realtime/runner.go (hunk 12 above)
Comment (Ananaskelly) [reply pkg/harness/openclaw/realtime/runner.go:389]: Yes, I understand the limitation, but OpenClaw currently does not emit a recording-level signal confirming that all submitted audio has been transcribed. The runner can only wait for additional transcript events during a quiet timeout after the latest observed event. We can extend this timeout, but I'm not sure if there is a reliable way to check whether we've received the full transcript. I've added comments about this in docs/voice_mode.md.

**Review by chengzhi-lu — APPROVED** [review]

**Comment by chengzhi-lu** [comment]
@Ananaskelly Please resolve the conflicts.

**Comment by Ananaskelly** [comment]
The conflicts are resolved. @chengzhi-lu
